src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.cs

```diff
@@ -265,13 +266,28 @@ await TrySetDigestAuthToken(request, challenge.Credential, digestResponse, isPro
                         break;
 
                     case AuthenticationType.Basic:
-                        if (performedBasicPreauth)
+                        if (preAuthCredential != null)
                         {
                             if (NetEventSource.Log.IsEnabled())
                             {
                                 NetEventSource.AuthenticationError(authUri, $"Pre-authentication with {(isProxyAuth ? "proxy" : "server")} failed.");
                             }
-                            break;
+
+                            if (challenge.Credential == preAuthCredential)
+                            {
+                                // Pre auth failed, and user supplied credentials are still same, we can stop there.
+                                break;
+                            }
+
+                            // Pre-auth credentials have changed, remove old ones from cache and continue with new ones.
+                            lock (pool.PreAuthCredentials!)
+                            {
+                                if (NetEventSource.Log.IsEnabled())
+                                {
+                                    NetEventSource.Info(pool.PreAuthCredentials, $"Pre-authentication credentials changed, removing Basic credential uri={preAuthCredentialUri}, username={preAuthCredential.UserName}");
+                                }
+                                pool.PreAuthCredentials.Remove(preAuthCredentialUri!, BasicScheme);
+                            }
```

Comment:
While we lock the PreAuthCredentials the logic leading to this can run in parallel, right? Do we have concurrency problem when `Remove` can throw? Perhaps use `TryRemove`?

Reply:
The method calls `Dictionary<TKey, TValue>.Remove()` internally, which does not throw if the key does not exist in the first place.

Reply:
But considering parallel requests, since we drop the lock between removing the creds and adding new ones, there may be some weird behavior if a parallel request starts when in-between. So I will move the removal to the same lock scope.